### PR TITLE
environment updates for the username userid split

### PR DESCRIPTION
**Pull request by butonic**
We updated the owncloud storage driver in reva to properly look up users by userid or username using the userprovider instead of taking the path segment as is. This requires the user service address as well as changing the default layout to the userid instead of the username. The latter is not considered a stable and persistent identifier.

requires a replace of reva for the PR https://github.com/cs3org/reva/pull/1033

### CODE REVIEW

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storagehome.go  3
- pkg/command/storagehome.go  1
- pkg/command/storageocdata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageoc.go  1
- pkg/flagset/storageeosdata.go  1
- pkg/flagset/storageocdata.go  2
- pkg/command/storagehomedata.go  2
- pkg/flagset/storageoc.go  2
         

Clones removed
==============
+ pkg/flagset/storageeos.go  -1
+ pkg/flagset/storageroot.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageoc.go  2
- pkg/flagset/storageeosdata.go  1
- pkg/command/storageoc.go  1
- pkg/command/storageocdata.go  2
- pkg/command/storagehomedata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storagehome.go  1
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storagehome.go  3
         

Clones removed
==============
+ pkg/flagset/storageeos.go  -1
+ pkg/flagset/storageroot.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storagehomedata.go  4
- pkg/flagset/storageocdata.go  2
- pkg/command/storageoc.go  1
- pkg/flagset/storageoc.go  2
- pkg/flagset/storagehome.go  3
- pkg/command/storagehomedata.go  2
- pkg/command/storagehome.go  1
- pkg/flagset/storageeosdata.go  1
- pkg/command/storageocdata.go  2
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageeosdata.go  1
- pkg/command/storagehomedata.go  2
- pkg/flagset/storageocdata.go  2
- pkg/command/storageoc.go  1
- pkg/flagset/storagehome.go  3
- pkg/flagset/storagehomedata.go  4
- pkg/command/storagehome.go  1
- pkg/flagset/storageoc.go  2
- pkg/command/storageocdata.go  2
         

Clones removed
==============
+ pkg/flagset/storageeos.go  -1
+ pkg/flagset/storageroot.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageeosdata.go  1
- pkg/command/storagehomedata.go  2
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageocdata.go  2
- pkg/command/storageoc.go  1
- pkg/flagset/storageoc.go  2
- pkg/command/storagehome.go  1
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageocdata.go  2
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/command/storagehomedata.go  2
- pkg/command/storageoc.go  1
- pkg/command/storageocdata.go  2
- pkg/command/storagehome.go  1
- pkg/flagset/storagehomedata.go  4
- pkg/flagset/storageoc.go  2
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageeosdata.go  1
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by PVince81** [comment]
"/srv/app/testrunner/tests/acceptance/features/apiWebdavUpload1/uploadFile.feature:63" is not supposed to run with OC-Storage, and yet it does here https://github.com/owncloud/core/blob/reva-tests/tests/acceptance/features/apiWebdavUpload1/uploadFile.feature#L63

@individual-it test setup issue ?

**Comment by PVince81** [comment]
the test has these tags: @skipOnOcV10 @skipOnOcis-OC-Storage @issue-ocis-reva-265`
maybe missing @skipOnOcis or the exclusion files need update

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/command/storageoc.go  1
- pkg/command/storagehomedata.go  2
- pkg/command/storagehome.go  1
- pkg/flagset/storageoc.go  2
- pkg/flagset/storagehome.go  3
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageocdata.go  2
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storageeosdata.go  1
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by phil-davis** [comment]
There are no `skipOnOcV10` test scenarios any more in core. https://github.com/owncloud/core/pull/37773

I guess this PR needs a rebase to latest master here in `ocis-reva`

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/command/storagehome.go  1
- pkg/flagset/storageoc.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageocdata.go  2
- pkg/flagset/storageeosdata.go  1
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageocdata.go  2
- pkg/command/storagehomedata.go  2
- pkg/command/storageoc.go  1
         

Clones removed
==============
+ pkg/flagset/storageeos.go  -1
+ pkg/flagset/storageroot.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storagehomedata.go  2
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageoc.go  2
- pkg/command/storageoc.go  1
- pkg/command/storagehome.go  1
- pkg/command/storageocdata.go  2
- pkg/flagset/storageeosdata.go  1
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by PVince81** [comment]
reva updated to get the sharing fixes from https://github.com/owncloud/ocis-reva/issues/439

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/command/storageoc.go  1
- pkg/flagset/storagehome.go  3
- pkg/command/storagehomedata.go  2
- pkg/flagset/storageeosdata.go  1
- pkg/command/storagehome.go  1
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageocdata.go  2
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storageoc.go  2
         

Clones removed
==============
+ pkg/flagset/storageeos.go  -1
+ pkg/flagset/storageroot.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by PVince81** [comment]
merging blocked by unexpected failing tests:

- apiOcisSpecific/apiShareManagementBasic-createShare.feature:116
- apiOcisSpecific/apiShareManagementBasic-createShare.feature:117

not sure if these are legit failures or intended to skip ? @individual-it

**Comment by phil-davis** [comment]
Those are local test scenarios that (were) demonstrating a bug.
https://cloud.drone.io/owncloud/ocis-reva/897/2/7
```
  @issue-ocis-reva-372 @issue-ocis-reva-243
  Scenario Outline: sharing subfolder of already shared folder, GET result is correct                                      # /drone/src/tests/acceptance/features/apiOcisSpecific/apiShareManagementBasic-createShare.feature:88
    Given using OCS API version "<ocs_api_version>"                                                                        # FeatureContext::usingOcsApiVersion()
    And these users have been created with default attributes and without skeleton files:                                  # FeatureContext::theseUsersHaveBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()
      | username |
      | Brian    |
      | Carol    |
      | David    |
      | Emily    |
    And user "Alice" has created folder "/folder1"                                                                         # FeatureContext::userHasCreatedFolder()
    And user "Alice" has shared folder "/folder1" with user "Brian"                                                        # FeatureContext::userHasSharedFileWithUserUsingTheSharingApi()
    And user "Alice" has shared folder "/folder1" with user "Carol"                                                        # FeatureContext::userHasSharedFileWithUserUsingTheSharingApi()
    And user "Alice" has created folder "/folder1/folder2"                                                                 # FeatureContext::userHasCreatedFolder()
    And user "Alice" has shared folder "/folder1/folder2" with user "David"                                                # FeatureContext::userHasSharedFileWithUserUsingTheSharingApi()
    And user "Alice" has shared folder "/folder1/folder2" with user "Emily"                                                # FeatureContext::userHasSharedFileWithUserUsingTheSharingApi()
    When user "Alice" sends HTTP method "GET" to OCS API endpoint "/apps/files_sharing/api/v1/shares"                      # OCSContext::userSendsToOcsApiEndpoint()
    Then the OCS status code should be "<ocs_status_code>"                                                                 # OCSContext::theOCSStatusCodeShouldBe()
    And the HTTP status code should be "<http_status_code>"                                                                # FeatureContext::thenTheHTTPStatusCodeShouldBe()
    And user "Alice" sends HTTP method "GET" to OCS API endpoint "/apps/files_sharing/api/v1/shares?path=/folder1/folder2" # OCSContext::userSendsToOcsApiEndpoint()
    And the response should contain 2 entries                                                                              # FeatureContext::checkingTheResponseEntriesCount()
    And folder "/folder1" should not be included as path in the response                                                   # FeatureContext::checkSharedFileAsPathNotInResponse()
    And folder "/folder2" should be included as path in the response                                                       # FeatureContext::checkSharedFileAsPathInResponse()

    Examples:
      | ocs_api_version | http_status_code | ocs_status_code |
      | 1               | 200              | 996             |
        OCS status code is not the expected value 996 got 100
        Failed asserting that two strings are equal.
        --- Expected
        +++ Actual
        @@ @@
        -'996'
        +'100'
      | 2               | 500              | 996             |
        OCS status code is not the expected value 996 got 200
        Failed asserting that two strings are equal.
        --- Expected
        +++ Actual
        @@ @@
        -'996'
        +'200'
```

It looks like the codes are better. So stuff has been fixed in `cs3org/reva` and now we are importing the fixed code from there.

I hope that there are "unexpected passes" in the Core-API tests - but they seem to have passed as they are. So there is no failing scenario from there that is now passing.

So maybe the local scenario here is improved but not going to fully pass every step yet.

For that, adjust the test scenario in this PR to expect the better status codes.

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageeosdata.go  1
- pkg/flagset/storagehome.go  3
- pkg/command/storageoc.go  1
- pkg/command/storagehomedata.go  2
- pkg/command/storageocdata.go  2
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storagehome.go  1
- pkg/flagset/storageoc.go  2
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/command/storageocdata.go  2
- pkg/command/storagehomedata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/flagset/storageocdata.go  2
- pkg/flagset/storageoc.go  2
- pkg/command/storagehome.go  1
- pkg/command/storageoc.go  1
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageeosdata.go  1
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- pkg/flagset/storageoc.go  2
- pkg/command/storagehome.go  1
- pkg/flagset/storagehome.go  3
- pkg/flagset/storageeosdata.go  1
- pkg/command/storageocdata.go  2
- pkg/flagset/storagehomedata.go  4
- pkg/command/storageoc.go  1
- pkg/flagset/storageocdata.go  2
- pkg/command/storagehomedata.go  2
         

Clones removed
==============
+ pkg/flagset/storageroot.go  -2
+ pkg/flagset/storageeos.go  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-reva/pullRequest?prid=5945743&bid=19491584)

**Comment by phil-davis** [comment]
I have adjusted the test scenario to the improved status codes, so taht it now reads the same as in `cs3org/reva`.
CI should pass - someone can review, and IMO this can be merged.

File: pkg/flagset/storagehome.go [hunk 1]
```diff
@@ -329,10 +329,20 @@ func StorageHomeWithConfig(cfg *config.Config) []cli.Flag {
...
+			Usage:       `"layout of the users home dir path on disk, in addition to {{.Username}}, {{.Mail}}, {{.Id.OpaqueId}}, {{.Id.Idp}} also supports prefixing dirs: "{{substr 0 1 .Username}}/{{.Username}}" will turn "Einstein" into "Ei/Einstein" `,
 			EnvVars:     []string{"REVA_STORAGE_OWNCLOUD_LAYOUT"},
 			Destination: &cfg.Reva.Storages.OwnCloud.Layout,
 		},
+
+		// User provider
+
+		&cli.StringFlag{
+			Name:        "users-url",
+			Value:       "localhost:9144",
```
Comment (kulmann) [inline pkg/flagset/storagehome.go:342]: I remember people complaining here, that a URL should always have `http(s)://`. Please adjust. ;-)

File: pkg/flagset/storagehomedata.go [hunk 2]
```diff
@@ -318,5 +318,15 @@ func StorageHomeDataWithConfig(cfg *config.Config) []cli.Flag {
 			EnvVars:     []string{"REVA_GATEWAY_URL"},
 			Destination: &cfg.Reva.Gateway.URL,
 		},
+
+		// User provider
+
+		&cli.StringFlag{
+			Name:        "users-url",
+			Value:       "localhost:9144",
```
Comment (kulmann) [inline pkg/flagset/storagehomedata.go:326]: Please add `http(s)://` (see first comment)

File: pkg/flagset/storageoc.go [hunk 3]
```diff
@@ -327,10 +327,20 @@ func StorageOCWithConfig(cfg *config.Config) []cli.Flag {
...
+			Usage:       `"layout of the users home dir path on disk, in addition to {{.Username}}, {{.Mail}}, {{.Id.OpaqueId}}, {{.Id.Idp}} also supports prefixing dirs: "{{substr 0 1 .Username}}/{{.Username}}" will turn "Einstein" into "Ei/Einstein" `,
 			EnvVars:     []string{"REVA_STORAGE_OWNCLOUD_LAYOUT"},
 			Destination: &cfg.Reva.Storages.OwnCloud.Layout,
 		},
+
+		// User provider
+
+		&cli.StringFlag{
+			Name:        "users-url",
+			Value:       "localhost:9144",
```
Comment (kulmann) [inline pkg/flagset/storageoc.go:340]: Please add `http(s)://` (see first comment)

File: pkg/flagset/storageocdata.go [hunk 4]
```diff
@@ -320,5 +320,15 @@ func StorageOCDataWithConfig(cfg *config.Config) []cli.Flag {
 			EnvVars:     []string{"REVA_GATEWAY_URL"},
 			Destination: &cfg.Reva.Gateway.URL,
 		},
+
+		// User provider
+
+		&cli.StringFlag{
+			Name:        "users-url",
+			Value:       "localhost:9144",
```
Comment (kulmann) [inline pkg/flagset/storageocdata.go:328]: Please add `http(s)://` (see first comment)

**Comment by PVince81** [comment]
I started adjusting locally and saw that other unrelated URLs are also not prefixed. I'd rather not risk more side effects now and leave it be, unless @butonic thinks it is safe to adjust them all...

**Comment by PVince81** [comment]
I've raised https://github.com/owncloud/ocis-reva/issues/440 for addressing the https thing

**Pull request merged**